certora/DSPC.spec

```diff
@@ -0,0 +1,376 @@
+    assert what != TOC() => tocAfter == tocBefore, "file did keep unchanged toc";
+}
+
+// Verify revert rules on file for global parameters
+rule file_global_revert(bytes32 what, uint256 data) {
+    env e;
+
+    mathint wardsSender = wards(e.msg.sender);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = wardsSender != 1;
+    bool revert3 = what != BAD() && what != TAU() && what != TOC();
+    bool revert4 = what == BAD() && to_mathint(data) != 0 && to_mathint(data) != 1;
+    bool revert5 = what == TAU() && to_mathint(data) > max_uint64;
+    bool revert6 = what == TOC() && to_mathint(data) > max_uint128;
+
+    file@withrevert(e, what, data);
+
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
+                            revert4 || revert5 || revert6,
+                            "file revert rules failed";
+}
+
+// Verify correct storage changes for non-reverting file for individual rate parameters
+rule file_per_id(bytes32 id, bytes32 what, uint256 data) {
+    env e;
+
+    mathint minBefore; mathint maxBefore; mathint stepBefore;
+    minBefore, maxBefore, stepBefore = cfgs(id);
+
+    file(e, id, what, data);
+
+    mathint minAfter; mathint maxAfter; mathint stepAfter;
+    minAfter, maxAfter, stepAfter = cfgs(id);
+
+    assert what == MIN() => minAfter == to_mathint(data), "file did not set min";
+    assert what != MIN() => minAfter == minBefore, "file did keep unchanged min";
+    assert what == MAX() => maxAfter == to_mathint(data), "file did not set max";
+    assert what != MAX() => maxAfter == maxBefore, "file did keep unchanged max";
+    assert what == STEP() => stepAfter == to_mathint(data), "file did not set step";
+    assert what != STEP() => stepAfter == stepBefore, "file did keep unchanged step";
+}
+
+// Verify revert rules on file for individual rate parameters
+rule file_per_id_revert(bytes32 id, bytes32 what, uint256 data) {
+    env e;
+
+    mathint wardsSender = wards(e.msg.sender);
+    mathint minBefore; mathint maxBefore; mathint stepBefore;
+    minBefore, maxBefore, stepBefore = cfgs(id);
+    mathint duty; mathint _rate;
+    duty, _rate = jug.ilks(id);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = wardsSender != 1;
+    bool revert3 = id != DSR() && id != SSR() && duty == 0;
+    bool revert4 = what != MIN() && what != MAX() && what != STEP();
+    bool revert5 = to_mathint(data) > max_uint16;
+    bool revert6 = what == MIN() && to_mathint(data) > maxBefore;
+    bool revert7 = what == MAX() && to_mathint(data) < minBefore;
+
+    file@withrevert(e, id, what, data);
+
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
+                            revert4 || revert5 || revert6 ||
+                            revert7,
+                            "file revert rules failed";
+}
+
+// Verify correct storage changes for non-reverting set for a single rate.
+rule set_single(bytes32 id, uint256 bps) {
```

Comment:
I've created [this PR](https://github.com/dewiz-xyz/dspc/pull/8) suggesting a `set` rule definition calling `DSPC::set` instead of using the `DSPCHarness`.